docs/proposals/20201123-tracing-ingestion-api-and-opentelemetry-collector.md

```diff
@@ -0,0 +1,87 @@
+
+* **Owners:**:
+    * `@pavolloffay`
+
+* **Related Tickets:**
+    * [observatorium/issues/442](https://github.com/observatorium/observatorium/issues/442)
+
+## Table of Contents
+
+- [TLDR](#tldr)
+- [Why](#why)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [How](#how)
+  * [Expose tracing ingestion API in the Observatorium API service](#expose-tracing-ingestion-api-in-the-observatorium-api-service)
+    + [Multitenancy](#multitenancy)
+  * [Add deployments manifests for OpenTelemetry collector](#add-deployments-manifests-for-opentelemetry-collector)
+- [Action Plan](#action-plan)
+- [References](#references)
+
+<small>
+
+<i>
+
+<a href="http://ecotrust-canada.github.io/markdown-toc/">
+Table of contents generated with markdown-toc
+</a>
+
+</i>
+
+</small>
+
+## TLDR
+
+This document proposes to expose tracing ingestion API  backed by the OpenTelemetry collector. This is a first step to support ingestion and storage for tracing data. The OpenTelemetry collector gives flexibility to forward tracing data to variety of tracing platforms, both on-prem or SaaS - Jaeger, Zipkin, Tempo, SigNoz or DataDog, Dynatrace etc.
+
+## Why
+
+At the moment Observatorium does not expose API to ingest distributed tracing data.
+
+## Goals
+
+* Expose API in the Observatorium API service to ingest tracing data.
+* Deploy a component that receives tracing data and is able to forward it to tracing platforms with storage.
+
+## Non-Goals
+
+* Store data in a persistent storage.
+* Provide read API for tracing data.
+
+## How
+
+The how is split into two sections:
+1. Expose tracing ingestion API
+2. Add deployment manifests for OpenTelemetry collector
+
+### Expose tracing ingestion API in the Observatorium API service
+
+There are a couple of open-source tracing protocols out there - Zipkin, Jaeger and OpenTelemetry. At the moment the OpenTelemetry seems to be the [most popular](https://opentelemetry.io/vendors/), and it is projected to have the biggest adoption in the future. Hence, using OpenTelemetry protocol seems to be the most appropriate choice.
+
+The OpenTelemetry protocol primarily supports gRPC for sending traces (OTLP gRPC) with proto encoding. The HTTP protocol with protobuf encoding is supported as well, but the JSON encoding is still in experimental mode. The majority of users use OTLP gRPC and the HTTP is used in environments where gRPC cannot be used (e.g. mobile clients). Because the Observatorium already supports HTTP and the final state is to support both OTLP HTTP and gRPC, the Observatorium could initially support HTTP and in parallel start working on gRPC.
+
+#### Multitenancy
+
+Multitenancy could be handled the same way as it is handled for metrics and logs. The API service could add an HTTP header or attribute/label to data. This label would then be used in the collector to identify the tenant. Then the collector could route the data to a specific exporter per tenant or export the data to platform/storage by using a single exporter.
```

Comment:
This also mentions collector. Why do we need any collector in Observatorium API? Why it cannot just be L7 proxy that terminates TLS and passes the grpc/http messages further?

Reply:
The question is then how the tenancy is planned to work. I believe this depends on the storage we use. Can tenancy works by just adding a new "Tag" to traces we can then isolate data from?

Observatorium API purpose is to allow this - is there a way to then query for traces from tenant X but not see traces from tenant Y while using our APIs? This is exactly something I think we missed in this proposal - what's the plan. Even if we want to skip tenancy for now - that should be a clear non goal then.

Reply:
> Can tenancy works by just adding a new "Tag" to traces we can then isolate data from?

This will not work with Jaeger at the moment.

> Observatorium API purpose is to allow this - is there a way to then query for traces from tenant X but not see traces from tenant Y while using our APIs? This is exactly something I think we missed in this proposal - what's the plan. Even if we want to skip tenancy for now - that should be a clear non goal then.

The proposal is simply about adding ingestion API and not query part.

Reply:
> This will not work with Jaeger at the moment.

So what will work? I don't see other solution proposed.

> The proposal is simply about adding ingestion API and not query part.

Ok, but ingesting dat without tenancy is blocking future use this data in multi-tenant world. I think we should at least discuss options here.

Reply:
I see you started great discussion on Jaeger itself, nice, great job! Looking though this productive  discussion now https://github.com/jaegertracing/jaeger/issues/3427

Reply:
I took a look at gRPC OTLP.  gRPC doesn't have paths so I imagine the tenant being a header.  This is not a problem for OTLP.
The OTLP collector has an extension that can do AuthN, but not AuthZ.  It is straightforward to have Observatorium listen for gRPC and handle `opentelemetry.proto.collector.trace.v1.TraceService` and the same code as Observatorium to do AuthN and RBAC, then forward the traces along.  Any objections to this approach?